src/sb/subscriber.rs

```diff
@@ -3,21 +3,144 @@
 // SPDX-FileContributor: Mathias Kraus
 
 use super::{
-    ffi::SubscriberStrongRef, sample::SampleReceiver, topic::SampleReceiverToken, SubscribeState,
-    Topic,
+    ffi, ffi::SubscriberStrongRef, sample::SampleReceiver, SubscribeState, SubscriberOptions,
 };
+use super::{mt, st};
+use crate::IceoryxError;
+use crate::QueueFullPolicy;
 
 use std::marker::PhantomData;
 
+pub struct SubscriberBuilder<'a, T> {
+    service: &'a str,
+    instance: &'a str,
+    event: &'a str,
+    options: SubscriberOptions,
+    phantom: PhantomData<T>,
+}
+
+impl<'a, T> SubscriberBuilder<'a, T> {
+    pub fn new(service: &'a str, instance: &'a str, event: &'a str) -> Self {
+        Self {
+            service,
+            instance,
+            event,
+            options: SubscriberOptions::default(),
+            phantom: PhantomData,
+        }
+    }
+
+    pub fn queue_capacity(mut self, queue_capacity: u64) -> Self {
+        self.options.queue_capacity = queue_capacity;
+        self
+    }
+
+    pub fn history_request(mut self, history_request: u64) -> Self {
+        self.options.history_request = history_request;
+        self
+    }
+
+    pub fn node_name(mut self, node_name: String) -> Self {
+        self.options.node_name = node_name;
+        self
+    }
+
+    pub fn queue_full_policy(mut self, queue_full_policy: QueueFullPolicy) -> Self {
+        self.options.queue_full_policy = queue_full_policy;
+        self
+    }
+
+    pub fn requires_publisher_history_support(
+        mut self,
+        requires_publisher_history_support: bool,
+    ) -> Self {
+        self.options.requires_publisher_history_support = requires_publisher_history_support;
+        self
+    }
+
+    pub fn create(mut self) -> Result<(st::Subscriber<T>, SampleReceiverToken), IceoryxError> {
+        self.options.subscribe_on_create = true;
+        let ffi_sub = ffi::Subscriber::new(self.service, self.instance, self.event, &self.options)
+            .ok_or(IceoryxError::SubscriberCreationFailed)?;
+
+        let subscriber = st::Subscriber {
+            ffi_sub: ffi::SubscriberRc::new(ffi_sub),
+            phantom: PhantomData,
+        };
+
+        Ok((subscriber, SampleReceiverToken {}))
+    }
+
+    pub fn create_mt(mut self) -> Result<(mt::Subscriber<T>, SampleReceiverToken), IceoryxError> {
+        self.options.subscribe_on_create = true;
+        let ffi_sub = ffi::Subscriber::new(self.service, self.instance, self.event, &self.options)
+            .ok_or(IceoryxError::SubscriberCreationFailed)?;
+
+        let subscriber = mt::Subscriber {
+            ffi_sub: ffi::SubscriberArc::new(ffi_sub),
+            phantom: PhantomData,
+        };
+
+        Ok((subscriber, SampleReceiverToken {}))
+    }
+
+    pub fn create_without_subscribe(mut self) -> Result<InactiveSubscriber<T>, IceoryxError> {
```

Comment:
I would also go back here and just use one create and make `subscribe_on_create` an option. If not you forgot `create_without_subscriber_mt` I think.

Reply:
Okay, what about turning the create into a generic where you state the compile time properties, threadsafe and subscribe on create, and the generic finds the most fitting type and returns it?
Something like:
```rust
    pub fn new() -> InactiveSubscriber<Self> {
        SubscriberBuilder::<Self>::new("Introspection", "RouDi_ID", "MemPool")
            .queue_capacity(1)
            .history_request(1)
            .build()
            .create<threadSafe::Nope, subscription::OnCreate>()
```

But if this does not make sense I think your approach is the most reasonable one.

Reply:
This depends on the specialization feature which is not yet stabilized.